articles/115_topic_and_service_name_mapping.md

```diff
@@ -0,0 +1,313 @@
+On systems where the implementation is handled for us by DDS (like with Connext), we use `rS_` as the common prefix.
+
+Here is a non-exhaustive list of prefixes:
+
+| ROS Subsystem        | Prefix |
+|----------------------|--------|
+| ROS Topics           | rt_    |
+| ROS Service Request  | rq_    |
+| ROS Service Response | rs_    |
+| ROS Service          | rS_    |
+| ROS Parameter        | rp_    |
+| ROS Action           | ra_    |
+
+### Substitution of the Namespace Delimiter
+
+The namespace delimiter in ROS 2 topic and service names, a forward slash (`/`), will be replaced with double underscores (`__`).
+The leading forward slash (`/`) is not replaced with double underscores (`__`), and is replaced instead by the ROS specific name prefix.
+
+### ROS to DDS Name Conversion Examples
+
+Here are some examples of translations between ROS topic names and DDS topic names:
+
+| ROS Name          | Subsystem | DDS Name              |
+|-------------------|-----------|-----------------------|
+| `/foo`            | Topic     | `rt_foo`              |
+| `/foo/bar`        | Topic     | `rt_foo__bar`         |
+| `/_foo/bar`       | Topic     | `rt__foo__bar`        |
+| `/foo/_bar`       | Topic     | `rt_foo___bar`        |
+| `/_foo/_bar`      | Topic     | `rt__foo___bar`       |
+| `/_foo/_bar/_baz` | Topic     | `rt__foo___bar___baz` |
+| `/foo/_/bar`      | Topic     | `rt_foo_____bar`      |
+
+## Concerns/Alternatives
+
+This section lists concerns about the proposed design and alternatives that were considered.
+
+### Alternative ROS to DDS Mapping
+
+There was some discussion of alternatives and concerns with respect to the ROS -> DDS translation.
+
+#### Capital Letter Substitution Alternative
+
+Since the forward slash (`/`) is replaced with double underscores (`__`) when translating to DDS from ROS topic names, two characters out of the 256 character limit are lost with each additional namespace.
+One proposed alternative was to add a constraint that ROS topic names could not use capital letters, and then capital letters could be used as the stand in for the forward slashes (`/`).
+
+Trade-offs:
+
+- Uses one less character per namespace and makes it easier to calculate the maximum length of a ROS topic or service name.
+- Prevents users from using capital letters in their names, which is constraining and might be a problem for backwards compatibility with ROS 1 topics and services.
+
+Rational:
+
+Preventing users from using capital letters was too constraining for the added benefit.
+
+#### Prefix with Double Underscores
+
+This alternative differs only in that it uses two underscores in the prefix, i.e. `rt__` rather than `rt_`.
+
+Trade-offs:
+
+- More consistent with replacement of other forward slashes (`/`)
+- Uses one more character
+
+Rational:
+
+Slight preference given to the shorter alternative.
```

Comment:
> It will greatly simplify the parsing process and visually it will also be consistent.

I agree it would be more visually consistent, but I take issue with both your and @dirk-thomas's characterization of how much more complicated one `_` rather than two `__` would be... I just don't think it's an issue at all. If you want the to split the header (`rX`) from the name, simply split it at the first `_` and prepend a namespace separator to the name. If you're creating the topic name, simply remove the leading separator, prepend an `_`, and then prepend the header (`rX`).

That being said, if no one agrees with me, then I'll update the document. Please everyone vote up/down on this proposed change.

> I would suggest that we reserve for ourselves several characters. Maybe 6 for the prefix. We will plan to use only 2 in our implementations. And assuming we are happy in the future we can decrease our reserved characters to the user.

I'm not sure I agree because I can't imagine why we'd need more characters (but it's hard to prove a negative). Especially given that we can start using the second character as an op-code if we need more options and then we have around 30 unique values (a-z + 0-9 - the ones we've already reserved). That would be a hit to readability and obviousness, but it's also something I don't think we'll have to do.

Also, when considering reserving more characters, I think that we should assume that we'll never give the characters back, because I think it would be a painful process to do so. So much so that it would never be worth rolling out. It would basically be a wire level breaking change to change the prefix. I'm not arguing that we shouldn't reserve space because of this, but it's like a mail-in rebate to me: I just assume the price does not include the rebate and that I'm not going to get it back, but then if I do it's icing on top.

> If we really wanted to preserve characters we'd want to use a single separator with an escape character.

What does this mean?

> The requirement that there's not trailing _ and leading numbers could make most separators only single width.

Can you give an example? I don't see how we could use those rules to implement a separator (I assume you mean the namespace delimiter).